### PR TITLE
Feature/OnRCStatus notification

### PR DESCRIPTION
**Pull request by LitvinenkoIra**
### This implementation is based on [RPC design refactoring](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0168-rpc-design-refactoring.md), so this PR should be merged after https://github.com/smartdevicelink/sdl_core/pull/2250

Technical task : #1930 
This PR is **ready** for review.

This pull request contains the implementation of new OnRCStatus notification
(proposal: https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0106-remote-control-onRcStatus-notification.md)
For testing use the following test set:
https://github.com/smartdevicelink/sdl_atf_test_scripts/blob/feature/rc_OnRCStatus/test_sets/rc_OnRCStatus.txt

Pull request with ATF scripts: https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/1911

For manual testing use:
HMI from : smartdevicelink/sdl_hmi#52
mobile application SPT from HockeyApp 20180213-Nightly-Android

### Risk
This PR makes **[major]** API changes.
### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by LitvinenkoIra** [comment]
Rebuild required

File: src/components/application_manager/include/application_manager/message_helper.h [hunk 1]
```diff
@@ -71,6 +71,22 @@ typedef std::map<std::string, mobile_apis::VehicleDataType::eType> VehicleData;
  **/
 class MessageHelper {
  public:
+  /**
+ * @brief CreateNotification creates basic mobile notification smart object
```
Comment (yang1070) [inline src/components/application_manager/include/application_manager/message_helper.h]: Notificaiton => Notification

File: src/components/application_manager/include/application_manager/message_helper.h [hunk 2]
```diff
@@ -71,6 +71,22 @@ typedef std::map<std::string, mobile_apis::VehicleDataType::eType> VehicleData;
...
+ * @param function_id Notificaiton function ID
+ * @param app_id application to send notification
+ * @return basic mobile notification smart object
+ */
+  static smart_objects::SmartObjectSPtr CreateNotification(
+      mobile_apis::FunctionID::eType function_id, uint32_t app_id);
+  /**
+ * @brief CreateHMINotification creates basic hmi notification smart object
```
Comment (yang1070) [inline src/components/application_manager/include/application_manager/message_helper.h]: typo notification

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/rc_command_request.cc [hunk 3]
```diff
@@ -140,6 +140,14 @@ void RCCommandRequest::Run() {
     SendResponse(false, mobile_apis::Result::DISALLOWED, "");
     return;
   }
+  if (!resource_allocation_manager_.is_rc_enabled()) {
```
Comment (yang1070) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/rc_command_request.cc]:
```
+  if (!resource_allocation_manager_.is_rc_enabled()) {
+    LOG4CXX_WARN(logger_, "Remote control is disabled by user");
+    SetResourceState(ModuleType(), ResourceState::FREE);
+    SendResponse(false,
+                 mobile_apis::Result::USER_DISALLOWED,
+                 "Remote control is disabled by user");
+    return;
+  }
```

File: src/components/application_manager/src/message_helper/message_helper.cc [hunk 4]
```diff
@@ -280,6 +280,45 @@ static VehicleInfo_Requests ivi_subrequests[] = {
...
+MessageHelper::CreateHMINotification(hmi_apis::FunctionID::eType function_id) {
+  using smart_objects::SmartObject;
+  using smart_objects::SmartObjectSPtr;
+  using smart_objects::SmartType_Map;
+  LOG4CXX_AUTO_TRACE(logger_);
+  SmartObjectSPtr notification_ptr =
+      utils::MakeShared<SmartObject>(SmartType_Map);
+  SmartObject& notification = *notification_ptr;
+  notification[strings::params][strings::message_type] =
+      static_cast<int32_t>(application_manager::MessageType::kNotification);
```
Comment (yang1070) [inline src/components/application_manager/src/message_helper/message_helper.cc]: why line 290 does not need `application_manager::MessageType::`, but it needs here

File: src/components/application_manager/include/application_manager/message_helper.h [hunk 5]
```diff
@@ -71,6 +71,22 @@ typedef std::map<std::string, mobile_apis::VehicleDataType::eType> VehicleData;
...
+ * @param function_id Notificaiton function ID
```
Comment (yang1070) [inline src/components/application_manager/include/application_manager/message_helper.h]: typo: Notificaiton

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/resource_allocation_manager_impl.cc [hunk 6]
```diff
@@ -213,10 +215,113 @@ void ResourceAllocationManagerImpl::RemoveAppsSubscriptions(const Apps& apps) {
   }
 }
 
+template <typename EnumType>
+std::string EnumToString(EnumType enum_type) {
+  using smart_objects::EnumConversionHelper;
+  const char* str = 0;
```
Comment (yang1070) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/resource_allocation_manager_impl.cc]: when will be the new allocated string released?

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/resource_allocation_manager_impl.cc [hunk 7]
```diff
@@ -213,10 +215,113 @@ void ResourceAllocationManagerImpl::RemoveAppsSubscriptions(const Apps& apps) {
   }
 }
 
+template <typename EnumType>
+std::string EnumToString(EnumType enum_type) {
```
Comment (yang1070) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/resource_allocation_manager_impl.cc]: cannot find where it is used

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/rc_command_request.cc (hunk 3 above)
Comment (LitvinenkoIra) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/rc_command_request.cc]: @yang1070 Fixed 35bd41c

File: src/components/application_manager/src/message_helper/message_helper.cc (hunk 4 above)
Comment (LitvinenkoIra) [reply src/components/application_manager/src/message_helper/message_helper.cc]: @yang1070 here we also don't need this
Removed 9a0fa27

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/resource_allocation_manager_impl.cc (hunk 7 above)
Comment (LitvinenkoIra) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/resource_allocation_manager_impl.cc]: @yang1070 Removed 9a0fa27

**Comment by LitvinenkoIra** [comment]
@yang1070 Fixed all typos and fix comments after review 
9a0fa27

**Comment by yang1070** [comment]
@theresalech @LuxoftAKutsan 
This pull request contains the implementation of onRCStatus. notification
Ford approved this PR.
It shall be rebased to merge into develop branch after RPC refactoring work.

**Comment by LitvinenkoIra** [comment]
change base branch from feature/rc_refactoring (not actual branch) to feature/vehicle_info_plugin (branch that contains all changes regarding PRC refactoring )

**Comment by LitvinenkoIra** [comment]
@JackLivio  feature/rc_OnRCStatus  rebased on latest develop

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/resource_allocation_manager_impl.cc [hunk 8]
```diff
@@ -213,10 +215,103 @@ void ResourceAllocationManagerImpl::RemoveAppsSubscriptions(const Apps& apps) {
...
+  EnumType val;
+  EnumConversionHelper<EnumType>::StringToEnum(str, &val);
+  return val;
+}
+
+void ConstructOnRCStatusNotificationParams(
+    smart_objects::SmartObject& msg_params,
+    const std::map<std::string, uint32_t>& allocated_resources,
+    const std::vector<std::string>& supported_resources,
+    const u_int32_t app_id) {
```
Comment (JackLivio) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/resource_allocation_manager_impl.cc]: Shouldn't this be `uint32_t`?

Comment (LitvinenkoIra) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/resource_allocation_manager_impl.cc]: @JackLivio You're right. Fixed 4799472a4d34c1e947ba5b075d31923877295748

**Comment by LitvinenkoIra** [comment]
feature/rc_OnRCStatus rebased on latest develop (af97446)

**Pull request merged**